### PR TITLE
[Feature] Exclude edges in sample_neighbors

### PR DESCRIPTION
**Pull request by mszarma**
## Description
Extending sample_neighbors API to support
exclude_edges parameter.

exclude_edges support tensor and dict data
Feature enable excluding certain edges
during neighborhood sampling
Exclude_edges contains EID's of edges
which will be excluded
during neighbor picking for seed nodes.

Added test case for heterograph and homograph
RFC issue id: https://github.com/dmlc/dgl/issues/2944

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
[Feature] Exclude edges in sample_neighbors
- Extending sample_neighbors API to support exclude_edges parameter.
 - Added test case for heterograph and homograph

### CODE REVIEW

File: python/dgl/sampling/neighbor.py [hunk 1]
```diff
@@ -52,6 +52,8 @@ def sample_neighbors(g, nodes, fanout, edge_dir='in', prob=None, replace=False,
         The features must be non-negative floats, and the sum of the features of
         inbound/outbound edges for every node must be positive (though they don't have
         to sum up to one).  Otherwise, the result will be undefined.
+    exclude_edges: tensor or dict
+        Edge IDs to exclude during sampling neighbors for the seed nodes.
```
Comment (zheng-da) [inline python/dgl/sampling/neighbor.py:56]: is it right that a tensor is for homogeneous graph and dict is for heterogeneous graph? if so, could you make it more explicit?

File: python/dgl/sampling/neighbor.py [hunk 2]
```diff
@@ -155,8 +181,21 @@ def sample_neighbors(g, nodes, fanout, edge_dir='in', prob=None, replace=False,
...
+        if not isinstance(exclude_edges, dict):
+            if len(g.etypes) > 1:
+                raise DGLError("Must specify etype type when the graph is not homogeneous.")
+            exclude_edges = {g.canonical_etypes[0] : exclude_edges}
+        exclude_edges = utils.prepare_tensor_dict(g, exclude_edges, 'edges')
+        for etype in g.canonical_etypes:
+            if etype in exclude_edges:
+                excluded_edges_all_t.append(F.to_dgl_nd(exclude_edges[etype]))
+            else:
+                excluded_edges_all_t.append(nd.array([], ctx=nd.cpu()))
```
Comment (zheng-da) [inline python/dgl/sampling/neighbor.py:195]: does it generate a right list of excluded edges? the list should be indexed by edge type IDs, but you iterate over the edge type based its order in the dict. There is no guarantee that the order is the same as indicated by edge type IDs.

File: src/array/cpu/coo_remove.cc [hunk 3]
```diff
@@ -81,9 +81,9 @@ COOMatrix COORemove(COOMatrix coo, IdArray entries) {
     return coo;
 
   std::vector<IdType> new_rows, new_cols, new_eids;
-  new_rows.reserve(nnz - n_entries);
-  new_cols.reserve(nnz - n_entries);
-  new_eids.reserve(nnz - n_entries);
+  new_rows.reserve(nnz > n_entries ? nnz - n_entries : nnz);
```
Comment (zheng-da) [inline src/array/cpu/coo_remove.cc:84]: if nnz < n_entries, return nnz? why?

File: src/graph/sampling/neighbor/neighbor.cc [hunk 4]
```diff
@@ -90,6 +91,8 @@ HeteroSubgraph SampleNeighbors(
         default:
           LOG(FATAL) << "Unsupported sparse format.";
       }
+      if (!exclude_edges.empty() && exclude_edges[etype].GetSize() != 0)
+        sampled_coo = COORemove(sampled_coo, exclude_edges[etype]);
```
Comment (zheng-da) [inline src/graph/sampling/neighbor/neighbor.cc:95]: what is the complexity of using this function? `SampleNeighbors` runs on the CSC format. adding this function requires to generate the COO format?

**Comment by zheng-da** [comment]
did you benchmark how fast is this implementation?

**Comment by mszarma** [comment]
> did you benchmark how fast is this implementation?

I benchmarked second version on synthetic bench, here are some examples of the results:

OLD = remove_edges + neighbor_sampling call
NEW = neighbor_sampling with excluded edges call
==================================================
old=0.0427086 , new=0.0160954 , diff=0.0266132 seconds, diff=62.31 percentage , faster

GRAPH num_nodes: 1019999  num_edges: 29855 
sampled_nodes_amount: 174  fanout: 3 excluded edges amount: 168 
SUBGRAPH num_nodes: 1019999  num_edges: 231 
==================================================
==================================================
old=12.5179353 , new=6.0757782 , diff=6.4421570 seconds, diff=51.46 percentage , faster

GRAPH num_nodes: 80002000  num_edges: 239989054 
sampled_nodes_amount: 689396  fanout: 6 excluded edges amount: 1990128 
SUBGRAPH num_nodes: 80002000  num_edges: 4136376
==================================================
==================================================
old=8.8966670 , new=6.6231253 , diff=2.2735417 seconds, diff=25.55 percentage , faster

GRAPH num_nodes: 80020000  num_edges: 239989189 
sampled_nodes_amount: 94441  fanout: 6 excluded edges amount: 76725 
SUBGRAPH num_nodes: 80020000  num_edges: 566646
==================================================
==================================================
old=0.0245998 , new=0.0203795 , diff=0.0042202 seconds, diff=17.16 percentage , faster

GRAPH num_nodes: 299992  num_edges: 149993 
sampled_nodes_amount: 8109  fanout: 178 excluded edges amount: 15545 
SUBGRAPH num_nodes: 299992  num_edges: 5420
==================================================
==================================================
old=0.0080178 , new=0.0049868 , diff=0.0030310 seconds, diff=37.80 percentage , faster

GRAPH num_nodes: 30000  num_edges: 30000 
sampled_nodes_amount: 7653  fanout: 9 excluded edges amount: 16394 
SUBGRAPH num_nodes: 30000  num_edges: 3677
==================================================

**Comment by mszarma** [comment]
> did you benchmark how fast is this implementation?

Yes, i bench second version on synthetic benchmark , here are some examples of results:
OLD = remove_edges + neighbor_sampling call
NEW = neighbor_sampling with excluded edges call
==================================================
old=0.0038526 , new=0.0017021 , diff=0.0021505 seconds, diff=55.82 percentage , faster

GRAPH num_nodes: 2997  num_edges: 1909 
sampled_nodes_amount: 735  fanout: 151 excluded edges amount: 1165 
SUBGRAPH num_nodes: 2997  num_edges: 198
==================================================
==================================================
old=0.0061331 , new=0.0036194 , diff=0.0025136 seconds, diff=40.99 percentage , faster

GRAPH num_nodes: 29998  num_edges: 29797 
sampled_nodes_amount: 1126  fanout: 196 excluded edges amount: 684 
SUBGRAPH num_nodes: 29998  num_edges: 1775
==================================================
==================================================
old=0.0427086 , new=0.0160954 , diff=0.0266132 seconds, diff=62.31 percentage , faster

GRAPH num_nodes: 1019999  num_edges: 29855 
sampled_nodes_amount: 174  fanout: 3 excluded edges amount: 168 
SUBGRAPH num_nodes: 1019999  num_edges: 231 
==================================================

==================================================
old=12.5179353 , new=6.0757782 , diff=6.4421570 seconds, diff=51.46 percentage , faster

GRAPH num_nodes: 80002000  num_edges: 239989054 
sampled_nodes_amount: 689396  fanout: 6 excluded edges amount: 1990128 
SUBGRAPH num_nodes: 80002000  num_edges: 4136376
==================================================
==================================================
old=8.8966670 , new=6.6231253 , diff=2.2735417 seconds, diff=25.55 percentage , faster

GRAPH num_nodes: 80020000  num_edges: 239989189 
sampled_nodes_amount: 94441  fanout: 6 excluded edges amount: 76725 
SUBGRAPH num_nodes: 80020000  num_edges: 566646
==================================================

**Comment by zheng-da** [comment]
how did you benchmark it? i think you can just take one dataset (e.g., OGBN-products) and benchmark it with and without using this PR for edge sampling.

**Review by mszarma — COMMENTED** [review]
> how did you benchmark it? i think you can just take one dataset (e.g., OGBN-products) and benchmark it with and without using this PR for edge sampling.

@zheng-da i made a simple bench test - used API calls and compare it with one new call.
OLD = remove_edges + neighbor_sampling call
NEW = neighbor_sampling with excluded edges call

 I think to benchmark it like you suggest master vs PR it's not available out of the box. Probably afaiu, someone with good understanding of this GNN use case would need to help to adapt upper layers code/calls to make it possible. Who could help with that? Jenny?

File: src/graph/sampling/neighbor/neighbor.cc [hunk 5]
```diff
@@ -104,6 +95,39 @@ HeteroSubgraph SampleNeighbors(
...
+    << "Number of fanout values must match the number of edge types.";
+  CHECK_EQ(prob.size(), hg->NumEdgeTypes())
+    << "Number of probability tensors must match the number of edge types.";
+
+  if (!exclude_edges.empty()) {
+    std::vector<IdArray> remain_edges(hg->NumEdgeTypes());
+    for (dgl_type_t etype = 0; etype < hg->NumEdgeTypes(); ++etype) {
+     EdgeArray edges = hg->Edges(etype);
+     remain_edges[etype] =
+            edges.excludeCertainEids<decltype(hg->NumEdges(etype))>(exclude_edges[etype]);
```
Comment (BarclayII) [inline src/graph/sampling/neighbor/neighbor.cc:120]: `NumEdges` is always `int64_t`.  So this function will fail if the graph is `int32`.

You can dispatch to different primitive types using `ATEN_ID_TYPE_SWITCH` (similarly there are other switches).  It works like this:
```C++
ATEN_ID_TYPE_SWITCH(hg->DataType(), IdType, {
  remain_edges[etype] =
      edges.excludeCertainEids<IdType>(exclude_edges[etype]);
});
```

Comment (mszarma) [reply src/graph/sampling/neighbor/neighbor.cc]: Done in newest commit version.

File: src/graph/sampling/neighbor/neighbor.cc [hunk 6]
```diff
@@ -104,6 +95,41 @@ HeteroSubgraph SampleNeighbors(
...
+    << "Number of node ID tensors must match the number of node types.";
+  CHECK_EQ(fanouts.size(), hg->NumEdgeTypes())
+    << "Number of fanout values must match the number of edge types.";
+  CHECK_EQ(prob.size(), hg->NumEdgeTypes())
+    << "Number of probability tensors must match the number of edge types.";
+
+  if (!exclude_edges.empty()) {
+    std::vector<IdArray> remain_edges(hg->NumEdgeTypes());
+    for (dgl_type_t etype = 0; etype < hg->NumEdgeTypes(); ++etype) {
+     EdgeArray edges = hg->Edges(etype);
```
Comment (zheng-da) [inline src/graph/sampling/neighbor/neighbor.cc:118]: the indent is not correct here.

File: src/graph/sampling/neighbor/neighbor.cc [hunk 7]
```diff
@@ -104,6 +95,41 @@ HeteroSubgraph SampleNeighbors(
   return ret;
 }
 
+HeteroSubgraph SampleNeighbors(
```
Comment (zheng-da) [inline src/graph/sampling/neighbor/neighbor.cc:98]: this implementation is not efficient on a large graph. you are remove the edges from the input graph. you should remove edges from the sampled subgraph.

**Comment by zheng-da** [comment]
you need to benchmark your implementation on a large graph.

**Comment by zheng-da** [comment]
you can take a look at the implementation of `EdgeDataLoader`: https://doc.dgl.ai/api/python/dgl.dataloading.html#dgl.dataloading.pytorch.EdgeDataLoader
it excludes certain edges.

File: python/dgl/dataloading/dataloader.py [hunk 8]
```diff
@@ -248,33 +253,8 @@ def sample_blocks(self, g, seed_nodes, exclude_eids=None):
...
-                    # so I need to keep track of the edge IDs left one by one.
-                    new_eids = parent_eids.copy()
-                    for k, v in located_eids.items():
-                        if len(v) > 0:
-                            frontier = transform.remove_edges(
-                                frontier, v, etype=k, store_ids=True)
-                            new_eids[k] = F.gather_row(parent_eids[k], frontier.edges[k].data[EID])
-                    frontier.edata[EID] = new_eids
+
+            frontier = self.sample_frontier(block_id, g, seed_nodes_in, exclude_eids=exclude_eids)
```
Comment (BarclayII) [inline python/dgl/dataloading/dataloader.py:257]: This is moving the edge ID exclusion into `sample_frontier`.  The reason we have edge ID exclusion after `sample_frontier` is that we would like to let user write their own neighbor samplers without taking care of edge exclusion.  Removing it like this will potentially break existing code since the edges are no longer automatically removed and users might do extra work.  @zheng-da do you think it's better to straight move it into `sample_frontier`?

File: python/dgl/sampling/neighbor.py (hunk 2 above)
Comment (BarclayII) [reply python/dgl/sampling/neighbor.py:328]: Is this resolved?  I saw it is now iterating over `g.canonical_etypes`.

File: src/graph/sampling/neighbor/neighbor.cc [hunk 9]
```diff
@@ -17,12 +17,57 @@ using namespace dgl::aten;
 namespace dgl {
 namespace sampling {
 
+HeteroSubgraph ExcludeCertainEdges(
```
Comment (zheng-da) [inline src/graph/sampling/neighbor/neighbor.cc:20]: how is this different from calling remove_edges in python?

Comment (zheng-da) [reply src/graph/sampling/neighbor/neighbor.cc:20]: actually, i'm kind of surprised that this solution has speedup over the existing solution.

File: python/dgl/dataloading/dataloader.py (hunk 8 above)
Comment (BarclayII) [reply python/dgl/dataloading/dataloader.py:257]: After discussion with Minjie we decide it's better to keep the original behavior so existing code will not break.  To have the best of both worlds we can add a flag in `BlockSampler`, say `exclude_edges_in_sample_frontier`.  When it is True, BlockSampler assumes that the edges must be excluded in `sample_frontier`, and `sample_blocks` no longer removes the edges.  Otherwise, it resorts to the current behavior where a frontier is first sampled then with edges removed.  The argument defaults to False, and could be set to True in `MultiLayerNeighborSampler` so `sample_neighbors` with `exclude_edges` can still be utilized.

**Comment by BarclayII** [comment]
Noticed another problem: now that we are putting `exclude_eids` in `sample_neighbors`, `exclude_eids_in_frontier` may not work by itself since we need to support the GPU version as well.  Will fix this later.

**Review by BarclayII — APPROVED** [review]
